### ISSUE TITLE
Can't run `make test-codeception`, `make nro-enable` or `make nro-test-codeception`

### ISSUE THREAD

**Issue by shawninder** [greenpeace/planet4-docker-compose#42]
I need to install p4 locally to do a tiny bit of development on the canada child theme. I followed the instructions exactly and although it mostly works (I can launch the website locally), the tests don't run. `make test-codeception` fails, saying `chown: changing ownership of '/app/source/public/wp-content/object-cache.php': No such file or directory`

The same message appears when running `make nro-enable` and `make nro-test-codeception`.

It seems your instructions are insufficient on a vanilla setup (I'm running a Macbook with OSX, nothing special).

---

On another note, have a look at this [documentation about P4 local development](https://planet4.greenpeace.org/create/tech/#5-local-development):

> We expect any self-respecting developer to be able to follow the instructions and setup their development environment. The instructions have been followed and worked for many developers until now.
> The P4 development team will not be able to help developers to build their own environment. (Blandly said: If you cannot build a development environment using the instructions, then you probably should not touching code until you manage it).

Ignoring the condescending part where you claim I'm not a self-respecting developer, a few questions still remain.
  1. If the instructions worked before for others, why are they not working for me?
  2. Where are the automated tests that prove these instructions should work on any system?
  3. Why is your team not able to help developers build their own environment? Any self-respecting developer should be able to provide working instructions to run their software, and support when those instructions fail (see what I did there?).
  4. If you're so confident about your instructions, why is helping out so difficult or time-consuming?

I guess the docs are clear and I shouldn't be expecting any help here, but if any of you self-respecting developers feel like coming back down to earth to mingle with us mortals, any help would be greatly appreciated.

**Comment by shawninder** [comment]
Another relevant observation:
```
$make php-shell
# cd tests/
# composer install
chown: changing ownership of '/app/source/public/wp-content/object-cache.php': No such file or directory
```

**Comment by comzeradd** [comment]
Hi @shawninder 

From the error message, I can assume that something went wrong with the installation and `object-cache.php` is missing from your local setup. Can you actually check that this file exists in the path mentioned, inside the persistence folder?

[TESTING](https://github.com/greenpeace/planet4-docker-compose/blob/master/TESTING.md) has some more detailed information about tests, and what are the steps to install test dependencies and run them. But as I said, the error you get means that something went wrong in a previous step.

I changed the wording on the linked handbook document, to make it more clear and inclusive. Apologies for that. The environment is currently being used by all P4 engineers, but also from NRO developers. That said, it's also expected that people run into issues sometimes. We always [invite](https://planet4.greenpeace.org/handbook/dev-starting-point/#communication-with-the-other-developers) developers from NROs to join our chat and get in touch with the dev community. Feel free to join. I'm not personally using OsX, but someone else maybe does.

**Comment by shawninder** [comment]
Yes, I saw the TESTING.md file, thanks. However, I think you're right about my problem happening before that file becomes relevant.

As for object-cache.php, here is what I see in the persistence folder:

```sh
$ ls -la persistence/app/public/wp-content/object-cache.php
lrwxrwxrwx  1 sinder  staff  63 10 Feb 14:53 persistence/app/public/wp-content/object-cache.php -> /app/source/public/wp-content/plugins/wp-redis/object-cache.php
```
As you can see, the file exists, but it's an alias or link pointing to `/app/source/public/wp-content/plugins/wp-redis/object-cache.php`

Now I'm not 100% sure about this next part, but I think the target file also exists:

```sh
$ ls -la persistence/app/public/wp-content/plugins/wp-redis/object-cache.php
-rw-r--r--@ 1 sinder  staff  41631 14 Dec  2018 persistence/app/public/wp-content/plugins/wp-redis/object-cache.php
```

Am I looking at the right file above?

As you can see it's a sizable file, and it contains what looks like 1377 lines of perfectly valid PHP code.

**Comment by shawninder** [comment]
@comzeradd I just made an account on chat.greenpeace.org using my Greenpeace email sinder@greenpeace.org
Would you mind approving it?

Thanks for your help

**Comment by comzeradd** [comment]
I have activated your account on RC.

**Comment by pablocubico** [comment]
Hey @shawninder I'm on a Mac but couldn't reproduce it yet, seems to work normally:

![image](https://user-images.githubusercontent.com/340766/74375222-b4c29f00-4dbe-11ea-9319-afb0059c121b.png)

I'll try destroying my environment and rebuilding from scratch.

**Comment by shawninder** [comment]
When I copy the original object-cache.php file and squash the symlink with it in hopes of fixing the problem, I can successfully run the tests, but one of the tests fails and complains about that file not existing. Here is the relevant test output:
```
1) 404PageCept: Check the 404 page looks nice
 Test  tests/acceptance/404PageCept.php
 Step  See in source "<h2>Sorry, we can't find that page!</h2>\r\n<h4>The page might be extinct...</h4>\r\nUse the search tool below and try your luck again."
 Fail  Failed asserting that  on page /thispagereallywillnotexist
--> <html><head></head><body><br>
<b>Warning</b>:  require_once(/app/source/public/wp-content/plugins/wp-redis/object-cache.php): failed to open stream: No such file or directory in <b>/app/source/public/wp-includes/load.php</b> on line <b>546</b><br>
<br>
<b>Fatal error</b>:  require_once(): Failed ope
[Content too long to display. See complete response in '/app/source/tests/_output/' directory]
--> contains "<h2>Sorry, we can't find that page!</h2> <h4>The page might be extinct...</h4> Use the search tool below and try your luck again.".

Scenario Steps:

 4. $I->seeInSource("<h2>Sorry, we can't find that page!</h2>\r\n<h4>The page might be extinct...</h4>\r\nUse the search tool below and try your luck again.") at tests/acceptance/404PageCept.php:10
 3. $I->amOnPage("/thispagereallywillnotexist") at tests/acceptance/404PageCept.php:8
 2. $I->grabFromDatabase("wp_options","option_value",{"option_name":"planet4_options"}) at tests/_support/AcceptanceTester.php:48
 1. $I->grabFromDatabase("wp_options","option_value",{"option_name":"planet4_options"}) at tests/_support/AcceptanceTester.php:48

Artifacts:

Png: /app/source/tests/_output/404PageCept.fail.png
Html: /app/source/tests/_output/404PageCept.fail.html
```

**Comment by shawninder** [comment]
I tried running `make run` and running the tests again, but I get the same thing. Now, looking at `/app/source/tests/_output/404PageCept.fail.html`, I see there are actually 2 errors there:
- require_once(): Failed opening required '/app/source/public/wp-content/plugins/wp-redis/object-cache.php'
- Uncaught Error: Call to undefined function wp_kses_normalize_entities() in /app/source/public/wp-includes/formatting.php:4316

Here is the full file:

**/app/source/tests/_output/404PageCept.fail.html**:
```html
$ make php-shell
root@21079f9ae2d8:/app/source# cat /app/source/tests/_output/404PageCept.fail.html
<html><head></head><body><br>
<b>Warning</b>:  require_once(/app/source/public/wp-content/plugins/wp-redis/object-cache.php): failed to open stream: No such file or directory in <b>/app/source/public/wp-includes/load.php</b> on line <b>546</b><br>
<br>
<b>Fatal error</b>:  require_once(): Failed opening required '/app/source/public/wp-content/plugins/wp-redis/object-cache.php' (include_path='.:/usr/share/php') in <b>/app/source/public/wp-includes/load.php</b> on line <b>546</b><br>
<br>
<b>Fatal error</b>:  Uncaught Error: Call to undefined function wp_kses_normalize_entities() in /app/source/public/wp-includes/formatting.php:4316
Stack trace:
#0 /app/source/public/wp-includes/class-wp-fatal-error-handler.php(190): esc_url('https://wordpre...')
#1 /app/source/public/wp-includes/class-wp-fatal-error-handler.php(147): WP_Fatal_Error_Handler-&gt;display_default_error_template(Array, false)
#2 /app/source/public/wp-includes/class-wp-fatal-error-handler.php(52): WP_Fatal_Error_Handler-&gt;display_error_template(Array, false)
#3 [internal function]: WP_Fatal_Error_Handler-&gt;handle()
#4 {main}
  thrown in <b>/app/source/public/wp-includes/formatting.php</b> on line <b>4316</b><br>
</body></html>
```

**Comment by hi-upchen** [comment]
I have totally the same error as yours 

```
chown: changing ownership of '/app/source/public/wp-content/object-cache.php': No such file or directory
```

I replace the link `wp-content/object-cache.ph` file with the true file. It can temporary solve this issue.

```
docker exec -it planet4dockercompose_php-fpm_1 /bin/bash
cd /app/source/public/wp-content/
rm object-cache.php
cp /app/source/public/wp-content/plugins/wp-redis/object-cache.php ./
```

**Comment by shawninder** [comment]
Ugh, now I get a different problem when I follow the exact same instructions from scratch (from scratch meaning I removed and deleted all docker instances and images and deleted the entire repository from my computer to start again from scratch).

This time the `make` step reports an error:

> ERROR: php-fpm is not running

Running make again fails saying:

> ERROR: for planet4dockercompose_php-fpm_1  UnixHTTPConnectionPool(host='localhost', port=None): Read timed out. (read timeout=60)
> 
> ERROR: for php-fpm  UnixHTTPConnectionPool(host='localhost', port=None): Read timed out. (read timeout=60)
> ERROR: An HTTP request took too long to complete. Retry with --verbose to obtain debug information.
> If you encounter this issue regularly because of slow network conditions, consider setting COMPOSE_HTTP_TIMEOUT to a higher value (current value: 60).

Third and subsequent attempts end with `make: *** [run] Error 1`.

Now, running the tests also produces a different error this time:
```
$ make test
chown: changing ownership of '/app/source/vendor/plugins/planet4-plugin-gutenberg-engagingnetworks/editorStyle.css': No such file or directory
make: *** [install-codeception] Error 1
``` 
 
So I tried the usual fix for tests, but that fails this time: 
```
$ make php-shell
# cd tests/
# composer install
chown: changing ownership of '/app/source/vendor/plugins/planet4-plugin-gutenberg-engagingnetworks/editorStyle.css': No such file or directory
```

So in other words, I'm unable to try your solution @hi-upchen, unfortunately, although it looks like something I tried earlier which failed, just using `docker exec -it planet4dockercompose_php-fpm_1 /bin/bash` instead of `make php-shell`

**Comment by lithrel** [comment]
Hi :wave:
I was _lucky_ enough to stumble on this one today, and I think I identified an issue.
@shawninder @hi-upchen if you still have this problem (or if it appears again some day), would you mind following this experiment ?

In your `planet4-docker-compose` folder, run
```console
> make run # if not running already
> eval $(make env)
> docker-compose exec php-fpm sh -c "chown app:root /app/bin/composer"
> docker-compose exec php-fpm sh -c "chown app:root /app/bin/composer.sh"
```
and in the `Makefile`, around line 434, add `--user app` so  that it looks like
```sh
	@echo "NRO dirname $(NRO_DIRNAME)"
	@docker-compose exec --user app php-fpm sh -c " \
```

Then relaunch the make command that was locked (`make nro-enable` or other).

My assumption (and what I've found in my case) is that in some context the composer wrapper 
(like [this one](https://github.com/greenpeace/planet4-docker/blob/fb8de926185101360e239812c94beaf8e939a269/src/planet-4-151612/php-fpm/bin/composer.sh)) takes an infinite amount of time to `chown` all the files, including node_modules content. By running directly composer with the proper user, we can bypass this chown process and still have files with the right owners (hopefully 😐). No more `chown` = no more chown error and no command waiting forever.

**Comment by shawninder** [comment]
At the moment, `make dev` fails:

> OCI runtime exec failed: exec failed: container_linux.go:349: starting container process caused "exec: \"/app/source/tasks/other/install-deps.sh\": stat /app/source/tasks/other/install-deps.sh: no such file or directory": unknown
> make: *** [repos] Error 126

`make test-codeception`

> ERRORS!
> Tests: 40, Assertions: 149, Errors: 2, Failures: 12, Skipped: 3.

`make nro-enable` now works though!  And `make run` too. 
`eval $(make env)` works as well, and so does running `cd tests && composer install` in `make php-shell`. Basically, things have changed a lot since I last tried all this!

That being said, when I tried your instructions:

```sh
$ docker-compose exec php-fpm sh -C "chown app:root /app/bin/composer"
sh: 0: Can't open chown app:root /app/bin/composer
```

I"m not the best as I am super new with this code base and I don't have too much time to put on this, but if there's anything else I can do to help let me know.

**Comment by lithrel** [comment]
> That being said, when I tried your instructions:
> docker-compose exec php-fpm sh -C "chown app:root /app/bin/composer"

It's because of uppercase option `-C`, you need to use lowercase `-c` :)
`docker-compose exec php-fpm sh -c "chown app:root /app/bin/composer"`

> I"m not the best as I am super new with this code base and I don't have too much time to put on this, but if there's anything else I can do to help let me know.

No worries ! I was posting in case you needed this now; I'll continue digging on the issue during the following weeks and try to find a proper fix. Thanks for testing :)

**Comment by lithrel** [comment]
Issue was closed by https://github.com/greenpeace/planet4-docker/pull/48 and https://github.com/greenpeace/planet4-docker/pull/47.  
  
This issue might reappear in some cases if you do a `make stop && make start`, the PR https://github.com/greenpeace/planet4-docker-compose/pull/49 will try to solve this specific problem.
